sklearn/metrics/_parallel_reductions.pyx

```diff
@@ -721,3 +748,154 @@ cdef class FastSquaredEuclideanArgKmin(ArgKmin):
+        ):
+        return RadiusNeighborhood(X=X, Y=Y,
+                       distance_metric=DistanceMetric.get_metric(metric, **metric_kwargs),
+                       radius=radius,
+                       chunk_size=chunk_size)
+
+    def __init__(self,
+                 X,
+                 Y,
+                 DistanceMetric distance_metric,
+                 DTYPE_t radius,
+                 ITYPE_t chunk_size = CHUNK_SIZE,
+    ):
+        PairwiseDistancesReduction.__init__(self, X, Y, distance_metric, chunk_size)
+
+        self.radius = radius
+        self.sort_results = False
+
+    cdef int _reduce_on_chunks(self,
+        const DTYPE_t[:, ::1] X,
+        const DTYPE_t[:, ::1] Y,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil except -1:
+        cdef:
+            ITYPE_t i, j
+            const DTYPE_t[:, ::1] X_c = X[X_start:X_end, :]
+            const DTYPE_t[:, ::1] Y_c = Y[Y_start:Y_end, :]
+            DTYPE_t dist_i_j
+
+        for i in range(X_c.shape[0]):
+            for j in range(Y_c.shape[0]):
+                dist_i_j = self.distance_metric.dist(
+                    &X_c[i, 0], &Y_c[j, 0], self.d)
+                if dist_i_j <= self.radius:
+                    deref(self.neigh_distances)[X_start + i].push_back(dist_i_j)
+                    deref(self.neigh_indices)[X_start + i].push_back(Y_start + j)
+
+        return 0
+
+    cdef void _on_X_prange_iter_finalize(self,
+            ITYPE_t thread_num,
+            ITYPE_t X_chunk_idx,
+            ITYPE_t X_start,
+            ITYPE_t X_end,
+    ) nogil:
+        cdef:
+            ITYPE_t idx, jdx
+
+        # Sorting neighbors for each query vector of X
+        if self.sort_results:
+            for idx in range(X_start, X_end):
+                _simultaneous_sort(
+                    deref(self.neigh_distances)[idx].data(),
+                    deref(self.neigh_indices)[idx].data(),
+                    deref(self.neigh_indices)[idx].size()
+                )
+
+    def compute(self,
+           str strategy = "auto",
+           bint return_distance = False,
+           bint sort_results = False
+    ):
+        # Won't be freed for reasons stated at their definition.
+        self.neigh_distances = new vector[vector[DTYPE_t]](self.n_X)
+        self.neigh_indices = new vector[vector[ITYPE_t]](self.n_X)
+
+        # TODO: setup thread local datastructures for supporting _parallel_on_Y
+        self.sort_results = sort_results
+        self._parallel_on_X()
+        return self._finalise_compute(return_distance)
+
+
+    def _finalise_compute(self,
+           bint return_distance
+    ):
+        if return_distance:
+            np_arrays_indices = []
+            np_arrays_distances = []
+
+            for i in range(self.n_X):
+                np_arrays_distances.append(buffer_to_numpy_array(deref(self.neigh_distances)[i].data(),
```

Comment:
You could even write a helper named `nested_vectors_to_numpy_arrays` to loop over the outer vectors and to also take care of the deref + `.data()`.

Reply:
```python
if return_distance:
    return (
        nested_vectors_to_numpy_arrays(self.neigh_distances),
        nested_vectors_to_numpy_arrays(self.neigh_indices),
    )
else:
   return nested_vectors_to_numpy_arrays(self.neigh_indices)
```

Reply:
And you also need to free `self.neigh_distances`  if `return_distance` is `False`.

Reply:
> You could even write a helper named nested_vectors_to_numpy_arrays to loop over the outer vectors and to also take care of the deref + .data().

This is what I did at first, but I got some weird problem with a missing symbol in the generated .so file (`_ZSt28__throw_bad_array_new_lengthv`) when typing it. This is related to numpy in some way, but I haven't yet be able to diagnose why. 

So yes, this is _highly ugly_ but as it worked and as this issue was capilotracting me, I decided to leave it as is for now to see if it was possible at least to have something working.

I will definitely change it. :+1: